client/deployments/client.go

```diff
@@ -157,6 +191,87 @@ func listArtifact(a artifactData, detailLevel int) {
 	fmt.Println("--------------------------------------------------------------------------------")
 }
 
+func (c *Client) DirectUpload(
+	artifactPath, token, url string,
+	noProgress bool,
+) error {
+	var bar *pb.ProgressBar
+
+	artifact, err := os.Open(artifactPath)
+	if err != nil {
+		return errors.Wrap(err, "Cannot read artifact file")
+	}
+
+	artifactStats, err := artifact.Stat()
+	if err != nil {
+		return errors.Wrap(err, "Cannot read artifact file stats")
+	}
+
+	// create pipe
+	pR, pW := io.Pipe()
+
+	writer := pW
+
+	req, err := http.NewRequest(http.MethodPut, url, pR)
+	if err != nil {
+		return errors.Wrap(err, "Cannot create request")
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.ContentLength = artifactStats.Size()
+
+	reqDump, _ := httputil.DumpRequest(req, false)
+	log.Verbf("sending request: \n%v", string(reqDump))
+
+	if !noProgress {
+		// create progress bar
+		bar = pb.New64(artifactStats.Size()).
+			Set(pb.Bytes, true).
+			SetRefreshRate(time.Millisecond * 100)
+		bar.Start()
+	}
+
+	go func() {
+		defer pW.Close()
+		defer artifact.Close()
+
+		var w io.Writer
+		w = writer
+		if !noProgress {
+			w = bar.NewProxyWriter(writer)
+		}
+
+		if _, err := io.Copy(w, artifact); err != nil {
+			writer.Close()
+			_ = pR.CloseWithError(err)
+			return
+		}
+
+		writer.Close()
+	}()
+
+	rsp, err := c.client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "POST /artifacts request failed")
+	}
+	defer rsp.Body.Close()
+	pR.Close()
+
+	rspDump, _ := httputil.DumpResponse(rsp, true)
+	log.Verbf("response: \n%v\n", string(rspDump))
+
+	if rsp.StatusCode != http.StatusOK {
+		if rsp.StatusCode == http.StatusUnauthorized {
+			log.Verbf("artifact upload to '%s' failed with status %d", req.Host, rsp.StatusCode)
+			return errors.New("Unauthorized. Please Login first")
+		}
+		return errors.New(
+			fmt.Sprintf("artifact upload to '%s' failed with status %d", req.Host, rsp.StatusCode),
+		)
+	}
```

Comment:
The error handling here must change. Any 2XX status code is a successful status code. AWS uses 200, Azure uses 201, but some day they might even change to 202 or 204. In any case, any status code >= 300 should be considered an error.
The error message for 401 responses does not make sense. Logging in does not help for a signed URL.

Reply:
fixed.